tiledb/ml/data_apis/pytorch_sparse.py

```diff
@@ -0,0 +1,122 @@
+
+        # Multiple workers - split workload
+        else:
+            per_worker = int(math.ceil(rows / worker_info.num_workers))
+            worker_id = worker_info.id
+            iter_start = worker_id * per_worker
+            iter_end = min(iter_start + per_worker, rows)
+
+        y_attr_name = self.y.schema.attr(0).name
+
+        x_shape = self.x.schema.domain.shape[1:]
+        y_shape = self.y.schema.domain.shape[1:]
+
+        # Loop over batches
+        for offset in range(iter_start, iter_end, self.batch_size):
+            # Yield the next training batch
+            y_batch = self.y[offset : offset + self.batch_size]
+            x_batch = self.x[offset : offset + self.batch_size]
+
+            # TODO: Both for dense case support multiple attributes
+            values_y = list(y_batch.items())[0][1]
+            values_x = list(x_batch.items())[0][1]
+
+            # Transform to TF COO format y data
+            y_data = np.array(values_y).ravel()
+            x_data = np.array(values_x).ravel()
+            if x_data.shape[0] != y_data.shape[0]:
+                raise ValueError(
+                    "X and Y should have the same number of rows, i.e., the 1st dimension "
+                    "of TileDB arrays X, Y should be of equal domain extent inside the batch"
+                )
+
+            x_coords = []
+            for i in range(0, self.x.schema.domain.ndim):
+                dim_name = self.x.schema.domain.dim(i).name
+                x_coords.append(np.array(x_batch[dim_name]))
+
+            # Normalise indices for torch.sparse.Tensor are batched with 0-index
+            x_coords[0] = np.vectorize(
+                lambda x: x - np.max(x_coords[0]) + self.batch_size - 1
+            )(x_coords[0])
+
+            # TODO: Sparse labels are not supported by Pytorch during this iteration for completeness
+            # we support the ingestion of sparseArray in labels, but loss and backward will fail due to
+            # SparseCPU backend
+            if self.iter_sparse_label:
+                y_coords = []
+                for i in range(0, self.y.schema.domain.ndim):
+                    dim_name = self.y.schema.domain.dim(i).name
+                    y_coords.append(np.array(y_batch[dim_name]))
+
+                # Normalise indices for torch.sparse.Tensor are batched with 0-index
+                y_coords[0] = np.vectorize(
+                    lambda y: y - np.max(y_coords[0]) + self.batch_size - 1
+                )(y_coords[0])
+
+                yield torch.sparse_coo_tensor(
+                    torch.tensor(list(zip(*x_coords))).t(),
+                    x_data,
+                    (self.batch_size, x_shape[0]),
+                    requires_grad=False,
+                ), torch.sparse_coo_tensor(
+                    torch.tensor(list(zip(*y_coords))).t(),
+                    y_data,
+                    (self.batch_size, y_shape[0]),
+                    requires_grad=False,
+                )
```

Comment:
1. This is a rather long expression, better break it to variables `x_tensor`, `y_tensor`
2. `x_tensor` is the same with the `else` clause; move it out of the condition and and `yield x_tensor, y_tensor` after the conditional.